### ISSUE TITLE
Bot not acknowledging when wish is rolled.

### ISSUE THREAD

**Issue by FxDvRV** [vivinano/MudaeAutoBot#50]
Hello, after testing whether or not the bot works with a minimum of 10 kakera, and a minimum of 50 kakera, the bot works perfectly fine. However after changing the minimum kakera to 400, and someone above that being rolled (also on the "namelist" and my wishlist) the bot did not acknowledge or make any attempt to claim it.
Any idea what went wrong?

**Comment by vivinano** [comment]
so there a claim window so if the bot has already successfully made a claim or a Claim not available message is shown then everything will be ignored

basically anytime you start the bot there is no claim window set but it the bot is running gets a claim window set then its gonna ignore things until you are able to claim again

**Comment by FxDvRV** [comment]
I had a claim at that time, if that's what you mean

**Comment by vivinano** [comment]
alright did it have wished by @FxDvRV  in the header ?

**Comment by vivinano** [comment]
did you roll and get an instant snipe or was it someone else roll and he instant sniped it ?
because there is a function that checks if its already claimed then it knows that it has already lost the race and ignores the snipe

**Comment by FxDvRV** [comment]
The thing is, I wasn't at my computer when it happened, which is the annoying part because I can't tell you any specifics. Also, isn't there a way to make it so the bot is always faster?

**Comment by vivinano** [comment]
![image](https://user-images.githubusercontent.com/33008397/120730207-f2b16e00-c4ae-11eb-9032-4c3a44974003.png)

so this is the part of the way that snipe time is calculated

but basically mudae is kinda slow and dumb and if it is to fast then mudae doesnt know that u added a reaction and then there also snipe protection so it has to be slow enough to bypass the snipe protection but fast enough to snipe the character 

there several factors I've tried to minimize alot of these factors but overal

there is program execution time >> Discord API interaction time >> Mudae Interaction time 

I only control the 1st  

![image](https://user-images.githubusercontent.com/33008397/120730484-94d15600-c4af-11eb-9340-9cf6d1ed670d.png)

there is a time.sleep at the end of that to kind of add a .5 global sleep mode to that in terms of instant snipe cases so it doesnt fall through

**Comment by FxDvRV** [comment]
What's the snipe protection time? Does the roller get some sort of priority or something? And what time would you recommend I use?

**Comment by vivinano** [comment]
well it depends on what togglesnipe you play on 

on certain modes there a window where only the roller can get it or wisher can get it 

most of that time management is taken out of your thoughts from using the $settings command so it takes all that optimization out of your mind 

![image](https://user-images.githubusercontent.com/33008397/120732445-53db4080-c4b3-11eb-9cce-3c7069d83405.png)


the only thing u can really change is the time.sleep(.5) just to take out that wait time but it also chances sniping to fast and causing mudae to say to fast

**Comment by FxDvRV** [comment]
Oh, well currently the server is set to default with no restrictions, so that shouldn't be an issue. What speed should I use for this setting?

**Comment by vivinano** [comment]
uh post the channeldata.txt and let me confirm that just in case

**Comment by FxDvRV** [comment]
Sure

_ __**Server Settings**__ _
(Server not premium)

· Prefix: **$** ($prefix)
· Lang: **en** ($lang)
· Claim reset: every **180** min. ($setclaim)
· Exact minute of the reset: xx:**23** ($setinterval)
· Reset shifted: by +**0** min. ($shifthour)
· Rolls per hour: **10** ($setrolls)
· Time before the claim reaction expires: **30** sec. ($settimer)
· Spawn rarity multiplicator for already claimed characters: **2** ($setrare)
· Server game mode: **1** ($gamemode)
· This channel instance: **1** ($channelinstance)
· Slash commands: enabled ($toggleslash)

· Ranking: enabled ($toggleclaimrank/$togglelikerank)
· Ranks displayed during rolls: claims only ($togglerolls)
· Hentai series: enabled ($togglehentai)
· Disturbing imagery series: enabled ($toggledisturbing)
· Rolls sniping: 0 ($togglesnipe)
· Kakera sniping: 0 ($togglekakerasnipe)
· Limit of characters per harem: **8100** ($haremlimit)
· Custom reactions: no ($claimreact)

· Kakera trading: **disabled** ($togglekakeratrade)
· Kakera calculation: claims and likes ranks (and number of claimed characters) ($togglekakeraclaim/$togglekakeralike)
· Kakera value displayed during rolls: enabled ($togglekakerarolls)
· $kakeraloot wishprotect: enabled ($togglewishprotect)

**Comment by vivinano** [comment]
alright so there only a .5 sleep from 

![image](https://user-images.githubusercontent.com/33008397/120733067-6d30bc80-c4b4-11eb-83f6-9aedfa64fa96.png)

you can either remove that line entirely and have it just deal with no sleep time and try to react super fast as  state with the comment 

Anarchy FTW

or keep that sleep and change the value lower like .1 or .05 its all personal preference

**Comment by FxDvRV** [comment]
Keeping the sleep time of 0.5, what would you say the lowest claim delay setting could be that will still allow for consistent claims (so not too fast that it won't register)

**Comment by vivinano** [comment]
I haven't tested it in a live server so its up to how competitive your server is to be honest you can try with maybe .1 and go lower or higher from there but idk mudae sometimes doesn't respond me and a friend had a small laugh over the bot being to fast for mudae at one point hence for the time.sleep

**Comment by FxDvRV** [comment]
Oh, 0.1 isn't too fast? I read somewhere on the main page that you should leave 1 second minimum or something. I'll try something like that for now though.

**Comment by vivinano** [comment]
that was for the older discord.py version cause it was async and was even faster lol

**Comment by FxDvRV** [comment]
Ah okay, i'll try it out then.

**Comment by FxDvRV** [comment]
Closing for now, I think that problem was a one time thing

**Comment by FxDvRV** [comment]
Okay it just happened again, someone with 537 kakera value (so above my minimum of 400) was ignored. They weren't on my wishlist or on the name list though. Can you tell me if these conditions are independent or not? So like, is it on the namelist AND 400 kakera minimum, or just anyone 400 minimum and anyone on the namelist?

**Comment by vivinano** [comment]
was in during a claim window ? so it is programmed to ignore even if condition is true IF the bot has already claimed something successfully

**Comment by FxDvRV** [comment]
Yup, I most definitely had a claim.

**Comment by vivinano** [comment]
when was it the bot last did a successful snipe ? and when was this character rolled ?

**Comment by FxDvRV** [comment]
The bot has successfully sniped 100% of the time in my test servers, with it following the rules I set. However in my main server it has failed to claim both times, however it successfully reacts to kakera 100% of the time on all servers.

> and when was this character rolled ?

I'm not sure what you're asking here sorry

**Comment by vivinano** [comment]
so does it print ? or doesnt print

**Comment by FxDvRV** [comment]
print...? I'm so sorry I'm not understanding what you mean

**Comment by vivinano** [comment]
does it print out "character found" or does it does it not print anything

**Comment by FxDvRV** [comment]
Nothing is printed in the console, it does however properly notice when someone is rolled by printing the "someone else rolled" message

**Comment by vivinano** [comment]
if it isnt printed then that means you are out of the claim window

**Comment by FxDvRV** [comment]
What do you mean by out of the claim window?

**Comment by vivinano** [comment]
![image](https://user-images.githubusercontent.com/33008397/120764170-e8f72d00-c4e5-11eb-9ad2-1a67568a9a0a.png)

so if nothing is printed then that means that it happened to be outside of the claim window so the bot thinks that is is unable to claim and wont attempt to "log" that something was found

I guess do a search on the server its been missing and do 

(Username) from:Mudae

**Comment by FxDvRV** [comment]
I 100% had a claim then, it reset just minutes prior to being rolled.

**Comment by vivinano** [comment]
send me the screen and feel and let me see the time stamps

**Comment by FxDvRV** [comment]
With a lowered minimum kakera, I've confirmed it works on the main server, something musth ave just happened those times.

**Comment by vivinano** [comment]
the only reason i can see is if there was a claim window set inaccurately i will leave it running on my server and see if it claims after then claim window and all that

**Comment by FxDvRV** [comment]
Okay thanks. How does it actually detect the claim window? It's not waiting 3 hours after I last claimed right? It actually knows when the reset is?

**Comment by vivinano** [comment]
no it does calculation for the time window

**Comment by vivinano** [comment]
so The only thing i can think of following my codes logic

is that something set your claim window improperly maybe your claim timer is set because of a kakera timer let me see the logic

**Comment by vivinano** [comment]
can u show me if you got a kakera wait timer by any chance ? before the character rolled ?

**Comment by FxDvRV** [comment]
I'll try find out if I had a kakera react available when the character was rolled then. (Assuming that's what you mean.)

**Comment by FxDvRV** [comment]
Yep, can confirm I had enough kakera power to do a react at the time the character was rolled.

**Comment by vivinano** [comment]
so there is nothing starting with ur username that has a timer ? that could possible set the window incorrectly ?

**Comment by FxDvRV** [comment]
Yup, nobody in the server has anything remotely close to my username, and at that moment I had both claim and kakera timers up guaranteed.

**Comment by vivinano** [comment]
which checks did it pass ? name and kakera right ?

**Comment by FxDvRV** [comment]
Sorry, what do you mean by that?

**Comment by vivinano** [comment]
so this character was in ur name list and kakera ?

**Comment by FxDvRV** [comment]
Out of the two that the bot missed, one of them was in name list and above kakera minimum, the second one was only above the kakera minimum.

**Comment by vivinano** [comment]
for the name one send me the picture of the character and its name in your name list and i guess the 2nd one also trying to see if there is something special about those that they failed

**Comment by FxDvRV** [comment]
I might have figured it out. During both times, I had the error where it says failed to find $settings for channel xxxxx (the channel where they were rolled). But then on the next line it said "Reading from File for channel xxxxx" so I assumed it was fine, but maybe it wasn't working after all and that's why?

**Comment by vivinano** [comment]
ah so if it fails to find any setting for a channel that channel is removed from the list of channels it monitors so yeah that means it not longer is looking at that channel 

if you search $togglehentai from:Mudae do you not have settings instance ?

**Comment by vivinano** [comment]
![image](https://user-images.githubusercontent.com/33008397/120789442-9f1b4080-c4ff-11eb-8ffc-9a70d583d838.png)

if something like this doesnt appear up then there is no setting then its get removed from the list as it doesnt have the variables to work off of. 

The file reading was added later so for those servers without settings you could set up the special instance of your server based off setting for other servers via a simple copy paste with a rename

**Comment by FxDvRV** [comment]
There's been several instances/$settings messages in that channel so it should be having no problem finding them. Someone said it might fail if there's a lot of activity in the channel but whenever I'm running the program that's not the case yet it still can't find the $settings like 40% of the time.

**Comment by vivinano** [comment]
![image](https://user-images.githubusercontent.com/33008397/120795641-bc540d00-c507-11eb-8eed-2d67f36b3f50.png)
 change the console line to True 
and if you get that error again could you take screenshots of the logs that get printed so I can send to discum's dev

**Comment by FxDvRV** [comment]
Okay sure, done. I'll let you know if it happens again?

**Comment by vivinano** [comment]
yeah there gonna be alot of log jarrgon so it might be annoying but if u are able to get the $setting then I could help the Dev improve their Search function

**Comment by FxDvRV** [comment]
Cool, also just making sure: discord doesn't have to be focused for the program to work right?

**Comment by vivinano** [comment]
nope I'm really against any kind of automation that requires your whole desktop to be locked up all of this can be ran on a raspberry pi mini computer if you wanted.

you can game you can watch videos you could uninstall discord it will still work

**Comment by FxDvRV** [comment]
Wait, you don't even need discord open? That's pretty damn cool.

**Comment by vivinano** [comment]
nope not at all. 

All of the function happens in the background only thing is that python stays running. 
but a cool thing with python is that you could change it into a  .pyw and the thing can run in the background as a service so it doesnt even need a spot on your task bar

**Comment by FxDvRV** [comment]
Haha that's awesome. Thanks for all the help then.

**Comment by FxDvRV** [comment]
Okay, it just happened again. Kakera value above my minimum of 400, claim ready, kakera react ready too if that matters. Person was not on my wishlist or namelist, but they were above the minimum of 400. Bot made no attempt to claim or any logging. The same $settings error did not occur. Any ideas what went wrong?

**Comment by vivinano** [comment]
Did you take a image for me ?

**Comment by FxDvRV** [comment]
I don't know how this will help you, but yeah here's an image. 
![image](https://user-images.githubusercontent.com/85224351/120805958-aa1ab300-c589-11eb-9b8c-af1e1b2a7e79.png)
(Not wished by me, not claimed by me)

**Comment by vivinano** [comment]
That still helps so I can see if it fitting the criteria I set uh do u have picture of the console ? Like the blue text that gets printed

**Comment by FxDvRV** [comment]
Just a whole bunch of this. 
![image](https://user-images.githubusercontent.com/85224351/120807936-d9caba80-c58b-11eb-8a27-0c8574eb9f1d.png)

**Comment by vivinano** [comment]
and those are 4 different unique channel id right ? no double entries

**Comment by vivinano** [comment]
and for that channel that stuff is missed can you post that channels .txt so i can verify

**Comment by FxDvRV** [comment]
Yup, that's right.

**Comment by FxDvRV** [comment]
> and for that channel that stuff is missed can you post that channels .txt so i can verify

Do you mean this?

_ __**Server Settings**__ _
(Server not premium)

· Prefix: **$** ($prefix)
· Lang: **en** ($lang)
· Claim reset: every **180** min. ($setclaim)
· Exact minute of the reset: xx:**11** ($setinterval)
· Reset shifted: by +**0** min. ($shifthour)
· Rolls per hour: **10** ($setrolls)
· Time before the claim reaction expires: **30** sec. ($settimer)
· Spawn rarity multiplicator for already claimed characters: **2** ($setrare)
· Server game mode: **1** ($gamemode)
· This channel instance: **1** ($channelinstance)
· Slash commands: enabled ($toggleslash)

· Ranking: enabled ($toggleclaimrank/$togglelikerank)
· Ranks displayed during rolls: **disabled** ($togglerolls)
· Hentai series: enabled ($togglehentai)
· Disturbing imagery series: enabled ($toggledisturbing)
· Rolls sniping: 0 ($togglesnipe)
· Kakera sniping: 0 ($togglekakerasnipe)
· Limit of characters per harem: **8100** ($haremlimit)
· Reacts: ****disabled**** ($togglereact)
· Custom reactions: no ($claimreact)

· Kakera trading: **disabled** ($togglekakeratrade)
· Kakera calculation: claims and likes ranks (and number of claimed characters) ($togglekakeraclaim/$togglekakeralike)
· Kakera value displayed during rolls: **disabled** ($togglekakerarolls)
· $kakeraloot wishprotect: enabled ($togglewishprotect)

**Comment by vivinano** [comment]
are each server the same ? cause they should be named by the channel id but yes
if this is the correct one matching the channel that is

**Comment by FxDvRV** [comment]
> are each server the same ? cause they should be named by the channel id but yes
> if this is the correct one matching the channel that is

Only one server matters, the rest are just testing ones with only myself in them. That's the channel.txt of the one where the bot isn't claiming correctly all the time.

**Comment by vivinano** [comment]
![image](https://user-images.githubusercontent.com/33008397/120810547-25904c00-c519-11eb-9622-28931b74bd54.png)

does this section of the code look like your code ?

**Comment by FxDvRV** [comment]
![image](https://user-images.githubusercontent.com/85224351/120810786-b0f7f480-c58e-11eb-80f2-8a8a2b697bc7.png)

**Comment by vivinano** [comment]
type `print(f"{waifu_wall.get(channelid,0)} is the Waifu wall for {channelid}")`
 `print(f"{next_claim(channelid)[0]} is the Next_Claim for {channelid}")`

![image](https://user-images.githubusercontent.com/33008397/120811763-5624b580-c51a-11eb-93fd-e6125d7df995.png)

right when me cursor is 

if you dont want to deal with all the crazy stuff just run it on the one live server for now to see the value of that 

basically as long as the waifu wall is not equal to the nextclaim then the claim wall is not set 

so it should constantly print the waifu wall / next_claim

ex: 155500 waifu wall
155500 next_claim 

(will ignore everything)

ex: 155500 waifu_wall
155501 next_claim

(will be in claim mode)

**Comment by FxDvRV** [comment]
![image](https://user-images.githubusercontent.com/85224351/120813179-01705180-c591-11eb-9349-4e9486b76f5b.png)
Is this right?

**Comment by vivinano** [comment]
no line it up with the IF

**Comment by FxDvRV** [comment]
Okay, I limited channels to the main server and ran it. Nothing different is happening in the console though.

**Comment by FxDvRV** [comment]
Nevermind, it looks like this.
![image](https://user-images.githubusercontent.com/85224351/120814277-126d9280-c592-11eb-829e-e0eba952da2d.png)

**Comment by vivinano** [comment]
okay that is correct so now at any point if the waifu wall changes then that means something has set the waifu wall hopefully its an actually claim if not we can locate what is causing the waifu wall to be set incorrectly

**Comment by FxDvRV** [comment]
Alright, what do I do now?

**Comment by vivinano** [comment]
yeah but also if someone claims it then it wont add the reaction 

the problem here tho is that its not even logging the character being detected so that has to something with the claim window

so basically what you do now is

you see the `waifu wall #` and the `next claim #`  if those 2 are equal like right now its `0 waifu wall` and `150261 next_claim` it means that it is monitoring and it will be in claim mode 

now if those number ever become `EQUAL` it is in `ignore mode` until the `next claim cycle`

**Comment by FxDvRV** [comment]
Okay, I understand. Currently they're still not equal and thus it's in claim mode. If this happens again I'll be sure to check whether or not this is still the case.

**Comment by vivinano** [comment]
To be clear if you do it a `VALID` claim the waifu wall will change they will be `EQUAL` until the next claim cycle

if you ever get a `FALSE` claim basically those 2 values becoming equal when they are not suppose to be

I need you to take a screenshot of the Logs and then go to discord and take a picture where you think the Error happen that set the `FALSE window` 
We then can examine it together and see what went wrong

**Comment by FxDvRV** [comment]
What logs do you want a screenshot of? 
![image](https://user-images.githubusercontent.com/85224351/120816420-100c3800-c594-11eb-9f54-a71718cfcf1f.png)
These are still the same, unless there are some other logs. 

What exactly do you want a picture of in discord? One of the missed claims or?

**Comment by vivinano** [comment]
so basically the screenshot your logs like that and something like this 
![image](https://user-images.githubusercontent.com/33008397/120817300-989cc100-c51f-11eb-89f8-95e92732c677.png)

so like oh hey i think that the last line set a claim window

**Comment by FxDvRV** [comment]
Uh well, these are what the logs look like, at the very end I claimed someone, and then rolled again. This seems about right no?

![image](https://user-images.githubusercontent.com/85224351/120818097-a2f9a200-c595-11eb-9b06-45c0a59c00e3.png)

![image](https://user-images.githubusercontent.com/85224351/120818352-e2c08980-c595-11eb-879a-e3cda52acae5.png)

Is this what you're asking for?

**Comment by vivinano** [comment]
okay now the bot is going to be in ignore mode until the next claim cycle

**Comment by vivinano** [comment]
so keep the bot running the next thing that pops up if those values remain the same then it will ignore then when the values become different then it will be back to claim mode

**Comment by FxDvRV** [comment]
Yeah, the values are still the same every time a roll happens.

Also, something weird with my most recent pic of the logs, it says "Someone else rolled"  twice but that never happened, it was me for all of them.

**Comment by vivinano** [comment]
hmm interesting i dont think that will be an issue `Yet` that is but we will see just keep it running basically what will happen is that 

waifu wall will stay the `same` but next_claim will `change` and when that happens then ur back into claim mode not sure when your next claim is up tho

**Comment by FxDvRV** [comment]
Sure, that's fine. Also, this error happened out of nowhere.
![image](https://user-images.githubusercontent.com/85224351/120822586-e9e99680-c599-11eb-88e1-a2a4e86c72e4.png)

**Comment by vivinano** [comment]
did a message get "edited" on that channel without an embed ?

**Comment by FxDvRV** [comment]
I looked at all the messages in that time, I can't really find any that seem out of the ordinary

**Comment by vivinano** [comment]
nothing that starts with your user name of has a timer ?

**Comment by FxDvRV** [comment]
Nope, also just then it claimed someone perfectly, someone whose conditions were similar to the other ones that it missed, so i think it's all good now.
